ast/src/parsed/mod.rs

```diff
@@ -496,3 +508,141 @@ impl<T> ArrayExpression<T> {
         }
     }
 }
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub enum TypeName<E> {
+    /// Boolean
+    Bool,
+    /// Integer (arbitrary precision)
+    Int,
+    /// Field element (unspecified field)
+    Fe,
+    /// String
+    String,
+    /// Column, shorthand for "int -> fe"
+    Col,
+    /// Algebraic expression
+    Expr,
+    /// Polynomial identity
+    Constr,
+    Array(ArrayTypeName<E>),
+    Tuple(TupleTypeName<E>),
+    Function(FunctionTypeName<E>),
+}
```

Comment:
Why can't we use the other Type type?

Reply:
You mean why do we have to distinguish between TypeName and Type? The main reason is that TypeName has an Expression as array length while Type has a concrete number. Also when we will have user-defined type, TypeName for a struct will just be a string and the Type will be the full struct type.